### PR TITLE
[inductor] Apply fx passes recursively to nested subgraphs

### PR DESCRIPTION
**Pull request by aakhundov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120777
* __->__ #120665

Summary: The current machinery of Inductor's `compile_fx` assumes that the incoming fx graph is flat. As a result, everything before `graph.run` is applied to the outermost graph. This assumption was valid before #119759, but now there is control flow bringing (arbitrarily deeply) nested fx subgraphs to `compile_fx`.

In this PR, we start extending the `compile_fx` machinery to deal with nested fx subgraphs. Namely, we recursively apply Inductor's `pre_grad`, `joint_graph`, and `post_grad` passes to the nested subgraphs in the incoming fx graph.

For the recursive application of the `pre_grad` passes (which require example inputs per subgraph), we don't pass example inputs for the nested subgraphs. A few different attempts to infer the latter via fake tensor prop has led to different side effects in the model. Therefore, to the nested subgraphs, we only apply a subset of `pre_grad` passes that doesn't require example inputs.

Test Plan:

```
$ python test/inductor/test_control_flow.py
...
----------------------------------------------------------------------
Ran 26 tests in 59.252s

OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames

### CODE REVIEW

File: torch/_inductor/fx_passes/pre_grad.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def lazy_init():
         from . import fb  # type: ignore[attr-defined]  # noqa: F401
 
 
-def pre_grad_passes(gm: torch.fx.GraphModule, example_inputs):
+def pre_grad_passes(gm: torch.fx.GraphModule, example_inputs=None):
```
Comment (eellison) [inline torch/_inductor/fx_passes/pre_grad.py:83]: Did it not work to construct the example_inputs from tensor meta? what were the failures ?

**Review by eellison — COMMENTED** [review]
looks good just one question

File: torch/_inductor/fx_passes/pre_grad.py (hunk 1 above)
Comment (aakhundov) [reply torch/_inductor/fx_passes/pre_grad.py:83]: `tensor_meta` doesn't include device. There is a device check here in the `pre_grad.py`:

https://github.com/pytorch/pytorch/blob/5b5bcf0470692c3bfb06b0459671b47f0c987f9e/torch/_inductor/fx_passes/pre_grad.py#L176

And we can't really extrapolate devices from the available `example_inputs` to the parent graph to the subgraph inputs.

**Review by eellison — APPROVED** [review]
Looks good, in a follow up we can extend support to pre-grad passes.

**Comment by aakhundov** [comment]
> in a follow up we can extend support to pre-grad passes

@eellison pre-grad passes are still applied to the outermost graph in their entirety. For the nested subgraphs, only those that don't require example inputs are applied (there are a few which don't). But yeah, would be great to enable all pre-grad passes for all subgraphs.

**Pull request closed without merging**